src/workerd/server/server.c++

```diff
@@ -1365,25 +1372,152 @@ public:
 
     kj::Own<WorkerInterface> getActor(kj::String id,
         IoChannelFactory::SubrequestMetadata metadata) {
+      kj::String idCopy = kj::str(id);
       return newPromisedWorkerInterface(service.waitUntilTasks,
-          getActorThenStartRequest(kj::mv(id), kj::mv(metadata)));
+          getActorThenStartRequest(kj::mv(id), kj::mv(metadata))).attach(
+              kj::defer([this, idCopy=kj::mv(idCopy)]() {
+            // Once this WorkerInterface is dropped, we need to run some cleanup tasks.
+            KJ_IF_SOME(entry, actors.find(idCopy)) {
+              entry->clientDropped();
+              if (!entry->hasClients()) {
+                // No clients remain connected, if no new clients connect before the expiration
+                // timeout, the actor will be removed by the cleanupLoop task.
+                entry->updateAccessTime();
+              }
+            }
+          }));
     }
 
     kj::Own<IoChannelFactory::ActorChannel> getActorChannel(Worker::Actor::Id id) {
       return kj::heap<ActorChannelImpl>(*this, kj::mv(id));
     }
 
+    // ActorContainer mostly serves as a wrapper around Worker::Actor.
+    // We use it to associate a HibernationManager with the Worker::Actor, since the
+    // Worker::Actor can be destroyed during periods of prolonged inactivity.
+    //
+    // We use a RequestTracker to track strong references to this ActorContainer's Worker::Actor.
+    // Once there are no Worker::Actor's left (excluding our own), `inactive()` is triggered and we
+    // initiate the eviction of the Durable Object. If no requests arrive in the next 10 seconds,
+    // the DO is evicted, otherwise we cancel the eviction task.
+    class ActorContainer final: public RequestTracker::Hooks {
+    public:
+      ActorContainer(kj::StringPtr key, ActorNamespace& parent, kj::Timer& timer)
+          : key(key),
+            tracker(kj::refcounted<RequestTracker>(*this)),
+            parent(parent),
+            timer(timer),
+            lastAccess(timer.now()) {}
+
+      ~ActorContainer() {
+        // Shutdown the tracker so we don't use active/inactive hooks anymore.
+        tracker->shutdown();
+
+        KJ_IF_SOME(a, actor) {
+          // Unknown broken reason.
+          auto reason = 0;
+          a->shutdown(reason);
+        }
+
+        // We need to make sure we're removed from the actors map.
+        parent.actors.erase(key);
+      }
+
+      void active() override {
+        // We're handling a new request, cancel the eviction promise.
+        shutdownTask = kj::none;
+        hasHibernated = false;
+      }
+
+      void inactive() override {
+        KJ_IF_SOME(a, actor) {
+          KJ_IF_SOME(m, a->getHibernationManager()) {
+            // The hibernation manager needs to survive actor eviction and be passed to the actor
+            // constructor next time we create it.
+            manager = kj::addRef(*static_cast<HibernationManagerImpl*>(&m));
```

Comment:
This downcast looks like it's making a lot of assumptions, but I see similar downcasts exist elsewhere in the code too. I guess `HibernationManagerImpl` is the only allowed implementation of `HibernationManager`? If that's the case, could we possibly remove the inheritance and just have a single concrete type called `HibernationManager`?

This cleanup could be in a separate PR since it's not really a new issue introduced here.